registry/client/repository.go

```diff
@@ -163,67 +158,156 @@ func (r *repository) Signatures() distribution.SignatureService {
 	}
 }
 
-type signatures struct {
-	manifests distribution.ManifestService
-}
-
-func (s *signatures) Get(dgst digest.Digest) ([][]byte, error) {
-	m, err := s.manifests.Get(dgst)
-	if err != nil {
-		return nil, err
+func (r *repository) Tags(ctx context.Context) distribution.TagService {
+	return &tags{
+		client:  r.client,
+		ub:      r.ub,
+		context: r.context,
+		name:    r.Name(),
 	}
-	return m.Signatures()
 }
 
-func (s *signatures) Put(dgst digest.Digest, signatures ...[]byte) error {
-	panic("not implemented")
-}
-
-type manifests struct {
-	name   string
-	ub     *v2.URLBuilder
-	client *http.Client
-	etags  map[string]string
+// tags implements remote tagging operations.
+type tags struct {
+	client  *http.Client
+	ub      *v2.URLBuilder
+	context context.Context
+	name    string
 }
 
-func (ms *manifests) Tags() ([]string, error) {
-	u, err := ms.ub.BuildTagsURL(ms.name)
+func (t *tags) Enumerate(ctx context.Context, refs []string, last string) (n int, err error) {
+	u, err := t.ub.BuildTagsURL(t.name)
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 
-	resp, err := ms.client.Get(u)
+	resp, err := t.client.Get(u)
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 	defer resp.Body.Close()
 
 	if SuccessStatus(resp.StatusCode) {
 		b, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return nil, err
+			return 0, err
 		}
 
 		tagsResponse := struct {
 			Tags []string `json:"tags"`
 		}{}
 		if err := json.Unmarshal(b, &tagsResponse); err != nil {
-			return nil, err
+			return 0, err
 		}
 
-		return tagsResponse.Tags, nil
+		copy(refs, tagsResponse.Tags)
+		if len(tagsResponse.Tags) > len(refs) {
+			// passed in refs was insufficient to capture all tags
+			return len(refs), nil
+		}
+		return len(tagsResponse.Tags), io.EOF
 	}
-	return nil, handleErrorResponse(resp)
+	return 0, handleErrorResponse(resp)
```

Comment:
Good point about last.

The API function as-is does not support internal expansion: the passed in slice can only be `copy`'d into.   This matches the signature of `Repositories`, so changing it would deviate from that.  I'm a torn on this: @stevvooe any ideas?